Separate tests on Windows CI to Admin-Required and Non-Admin-Required

vors opened the pull request:
Fixes #1486 
- Add 'RequireAdminOnWindows' tag to some tests suites
- Change install scope for Install-PackageProvider.Tests.ps1
  This helps avoid problem when run as non-elevated user.
- Restructure PowerShellGet.Tests.ps1 to enable 'RequireAdminOnWindows' tagging
- Add Start-PSPester -Unelevated
  That will enable running tests without Admin in CI
- Make Invoke-AppVeyorTest run elevated and non-elevated tests separately

Before merging, we would need to update documentation:
- [ ] About using `Start-PSPester -Unelevate`
- [x] About Tags that are used in our Pester tests (add `RequireAdminOnWindows`)

Commit:
Add 'RequireAdminOnWindows' tag to some tests suites

Commit:
Change install scope for Install-PackageProvider.Tests.ps1

This helps avoid problem when run as non-elevated user.

Commit:
Restructure PowerShellGet.Tests.ps1 to enable 'RequireAdminOnWindows' tagging

Commit:
Add Start-PSPester -Unelevated

That will enable running tests without Admin in CI

Commit:
Make Invoke-AppVeyorTest run elevated and non-elevated tests separately

- Add RequireAdminOnWindows to the list of blessed tags
- Fix a bug in Get-PesterTag where we accepted 'Slow' as a proper priority
- Fix missed comma in Start-PSPester parameters
- Add documentation about new Pester tag
- Fix finishing logic for Start-PSPester -Unelevate

msftclas commented:
Hi __@vors__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Sergei Vorobev). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

vors commented:
I observed that we get test failures because we run tests that require admin without admin.
Currently I didn't find to `Invoke-Pester` with tests with **all** tests.
Opened https://github.com/pester/Pester/issues/608 for discussion.

Commit:
Replace Transcription by *> redirection

due to the issue https://github.com/PowerShell/PowerShell/issues/2334

Commit:
Timezone tests require Admin on Windows

vors commented:
This is blocked on https://github.com/PowerShell/PowerShell/issues/2346
cc @mirichmo

JamesWTruher commented on build.psm1:
```diff
@@ -615,10 +663,40 @@ function Start-PSPester {
+                {
+                    $lines = Get-Content $outputBufferFilePath | Select-Object -Skip $currentLines
+                    $lines | Write-Host
+                    if ($lines | ? { $_ -eq '__UNELEVATED_TESTS_THE_END__'})
```
this looks like debug output - is it necessary?

vors replied:
No, that's how the parent process gets notification that the child process ended.

JamesWTruher replied:
ok

JamesWTruher commented on build.psm1:
```diff
@@ -615,10 +663,40 @@ function Start-PSPester {
-            & $powershell -noprofile -c $Command
+            if ($Unelevate)
+            {
+                Start-UnelevatedProcess -process $powershell -arguments @('-noprofile', '-c', $Command)
```
>art-UnelevatedProcess [](start = 18, length = 21)

it's too bad that runas doesn't give you _any_ useful info about the running process

vors replied:
Indeed.
I was trying to find a way to avoid what I'm doing here.

JamesWTruher commented on build.psm1:
```diff
@@ -615,10 +663,40 @@ function Start-PSPester {
+                    $lines | Write-Host
+                    if ($lines | ? { $_ -eq '__UNELEVATED_TESTS_THE_END__'})
+                    {
+                        break
```
>break [](start = 24, length = 5)

are you relying on CI to bail if things take too long?

vors replied:
yeah. Should I include an explicit timeout logic?

JamesWTruher replied:
possibly, i believe that appveyor has it's own time limitations on execution, it may be that nothing is needed here (or rather, any number you select will probably be wrong, up to the value that AppVeyor has)

vors replied:
Yeah, we control the timeout for all our project globally and it's currently 3 hours. I would leave it as is for now.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1:
```diff
@@ -25,6 +25,7 @@ Describe "New-Item" -Tags "CI" {
     $FullyQualifiedLink   = Join-Path -Path $tmpDirectory -ChildPath $testlink
 
+
     BeforeEach {
```
empty line above this why?

vors replied:
Artifacts from my experiments. Fixing.

vors commented on build.psm1:
```diff
@@ -568,16 +579,43 @@ function Start-PSPester {
+            throw '-Unelevate cannot be applied because the current user is not Administrator'
+        }
+
+        if (-not $PSBoundParameters.ContainsKey('RequireAdminOnWindows'))
```
This should be 
```
if (-not $PSBoundParameters.ContainsKey('ExcludeTag'))
```

JamesWTruher commented on tools/appveyor.psm1:
```diff
@@ -173,48 +173,63 @@ function Invoke-AppVeyorTest
-    Start-PSPester -bindir $env:CoreOutput -outputFile $testResultsFile @coreClrTestParams
-    Write-Host -Foreground Green 'Upload CoreCLR test results'
-    Update-AppVeyorTestResults -resultsFile $testResultsFile
+    Start-PSPester -bindir $env:CoreOutput -outputFile $testResultsNonAdminFile @coreClrTestParams
```
>coreClrTestParams [](start = 81, length = 17)

don't you need to exclude RequiresAdmin tag for non admin?

vors replied:
Great catch!

Commit:
Fix bug about Admin tests defaults Start-PSPester in build.psm1

Commit:
Fix appveyor test runner logic for Admin and non-Admin tests

Commit:
Make Registry.Tests.ps1 admin-aware

vors commented:
Also found #2366 during this work. Will include fix for it in the next iteration of the PR.

Commit:
Make Get-PesterTag allow only one scope tag (CI, Feature, Priority)

Commit:
Make LocalGroup tests admin-aware

Commit:
Make ExecutionPolicy.Tests.ps1 admin-aware

Commit:
Allow ExecutionPolicy tests to throw specific type of error

Split LocalMachine test case into two separate tests:
- One without overriden policy (throws exception)
- One with overriden policy

vors commented:
Alright, tests are green and now it's finally ready for the review.

@mirichmo , can you please take a look at c758b2e ?
@JamesWTruher can you take a look at the rest?

vors commented:
@mirichmo done

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1:
```diff
@@ -998,9 +1003,43 @@ ZoneId=$FileType
+            }
+
+            $errorId | Should Be 'ExecutionPolicyOverride,Microsoft.PowerShell.Commands.SetExecutionPolicyCommand'
+            Get-ExecutionPolicy -Scope LocalMachine | Should Be "ByPass"
```
>Get-ExecutionPolicy -Scope LocalMachine | Should Be "ByPass" [](start = 12, length = 60)

I don't quite understand this:
attempt to set the localmachine policy, which is supposed to fail
then check the localmachine policy to see if it is what you just tried to set it to?
Shouldn't you get the localmachine policy in the BeforeAll, store it, then check against that value (rather than assume that ByPass will be correct)

vors replied:
Nope, that's how PowerShell works. The full error message text is:
```

Set-ExecutionPolicy : Windows PowerShell updated your execution policy successfully, but the setting is overridden by
a policy defined at a more specific scope.  Due to the override, your shell will retain its current effective
execution policy of RemoteSigned. Type "Get-ExecutionPolicy -List" to view your execution policy settings. For more
information please see "Get-Help Set-ExecutionPolicy".

FullyQualifiedErrorId : ExecutionPolicyOverride,Microsoft.PowerShell.Commands.SetExecutionPolicyCommand
```

**Windows PowerShell updated your execution policy successfully, but the setting is overridden by
a policy defined at a more specific scope.**

vors replied:
I'm going to add comments about it in the code, because it's indeed confusing

Commit:
Add verbose comment about the execution policy test

'-Scope LocalMachine is Settable, but overridden'

mirichmo commented on test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1:
```diff
@@ -998,9 +1003,54 @@ ZoneId=$FileType
+                $errorId = $_.FullyQualifiedErrorId
+            }
+
+            $errorId | Should Be 'ExecutionPolicyOverride,Microsoft.PowerShell.Commands.SetExecutionPolicyCommand'
```
The guidance for FQEID validation is:
```
try {
    Set-ExecutionPolicy -Scope LocalMachine -ExecutionPolicy ByPass 
    throw "Expected exception: ExecutionPolicyOverride"
}
catch {
    $_.FullyQualifiedErrorId | Should Be 'ExecutionPolicyOverride,Microsoft.PowerShell.Commands.SetExecutionPolicyCommand' 

}
```
}

vors replied:
done

mirichmo commented on test/powershell/Modules/Microsoft.PowerShell.Security/ExecutionPolicy.Tests.ps1:
```diff
@@ -998,9 +1003,54 @@ ZoneId=$FileType
-        # This test requires Administrator privileges on Windows.
-        It "-Scope LocalMachine is Settable" {
+        It '-Scope LocalMachine is Settable' {
+            Set-ExecutionPolicy -Scope CurrentUser -ExecutionPolicy Undefined
```
Add a comment here explaining why this is necessary.

mirichmo reviewed (changes requested):
I'd like to see two minor modifications. Other than that it looks good.

mirichmo reviewed (approved):
The ExecutionPolicy changes LGTM

JamesWTruher commented:
:shipit:
